website/statistics.py

```diff
@@ -362,3 +515,101 @@ def get_general_class_stats(students):
+    if not quizzes_ran_per_level[level - 1]:
+        avg_quizzes_ran_per_level.append(0)
+    else:
+        quizzes_per_level = quizzes_ran_per_level[level - 1]
+        if len(quizzes_ran_per_level[level - 1]) == 1:
+            avg_quizzes_ran_per_level.append(quizzes_per_level[0])
+        if len(quizzes_ran_per_level[level - 1]) > 1:
+            average_quiz_score_per_level = 0
+            for index in range(1, len(quizzes_per_level) + 1):
+                average_quiz_score_per_level += quizzes_per_level[index - 1]
+
+            average_quiz_score_per_level /= (len(quizzes_ran_per_level[level - 1]))
+            avg_quizzes_ran_per_level.append(int(average_quiz_score_per_level))
+
+    return avg_quizzes_ran_per_level
+
+
+# count the number of programs that the student has run each level
+def calc_num_programs_per_level(programs_ran_per_level, success_runs_per_level, error_runs_per_level, level):
+    if len(programs_ran_per_level[level - 1]) != 0:
+        successful_runs = 0
+        unsuccessful_runs = 0
+        for program in programs_ran_per_level[level - 1]:
+            if str(program[1]) == "True":
+                unsuccessful_runs += 1
+            else:
+                successful_runs += 1
+        success_runs_per_level[level - 1] = successful_runs
+        error_runs_per_level[level - 1] = unsuccessful_runs
+    else:
+        success_runs_per_level[level - 1] = 0
+        error_runs_per_level[level - 1] = 0
+
+    return success_runs_per_level, error_runs_per_level
+
+
+def find_success_rate_overall(quizzes):
+    success_rate_overall = "-"
+    if len(quizzes) != 0:
+        num_finished_quizzes = 0
+        success_rate_overall = 0
+        for index in range(1, len(quizzes) + 1):
+            finished_quiz = quizzes[index - 1]['finished']
+            started_quiz = quizzes[index - 1]['started']
+
+            num_finished_quizzes += finished_quiz
+            success_rate_overall += finished_quiz / started_quiz * 100
+
+        success_rate_overall /= num_finished_quizzes
+        success_rate_overall = round(success_rate_overall, ndigits=0)
+    return success_rate_overall
+
+
+def calc_highest_success_rate(finished_quizzes, highest_level_quiz, quizzes):
+    success_rate_highest_level = '-'
+    if finished_quizzes:
+        highest_level_started = ([x.get("started") for x in quizzes if x.get("level") == highest_level_quiz])
+        highest_level_finished = ([x.get("finished") for x in quizzes if x.get("level") == highest_level_quiz])
+        success_rate_highest_level = (highest_level_finished[0] / highest_level_started[0] * 100)
+        success_rate_highest_level = round(success_rate_highest_level, ndigits=0)
+    return success_rate_highest_level
+
+
+def calc_average_quizzes(average_quizzes_ran_per_level):
+    average_quizzes = "-"
+    num_quizzes = 0
+    total_quiz_scores = 0
+    for i in range(1, len(average_quizzes_ran_per_level) + 1):
+        if average_quizzes_ran_per_level[i - 1] != 0:
```

Comment:
Why are you starting from 1 and then removing 1 in the index? Maybe there's a valid reason! It's just a little convoluted to me.

Reply:
Oh, I think that is by mistake. I changed it now.